BIP 53: Disallow 64-byte transactions

Christewart opened the pull request:
This BIP makes 64 byte bitcoin transactions serialized without the witness consensus invalid

Mailing list post: https://groups.google.com/g/bitcoindev/c/rf3QOlzg230/m/eOOC8pkOAgAJ

Delving discussion: https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710

Side note, wasn't sure how to best handle the 2-Bitcoin-Merkle.pdf as it isn't hosted anywhere easily accessible AFAICT? For now I just added it into the git repo. Lmk if you have other suggestions

murchandamus reviewed (commented):
I am a little surprised at this pull request, given the [frequent coverage](https://bitcoinops.org/en/topics/consensus-cleanup-soft-fork/) of Antoine’s [ongoing research](https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/5) on reviving the Great Consensus Cleanup that includes a similar fix. 
It seems better to me to package all four fixes in one soft fork. Was it overlooked that this work is being duplicated, or are you disagreeing with the direction of Antoine’s work and want to put forward an alternative? Could you elaborate why you decided to submit this?

Christewart commented:
> I am a little surprised at this pull request, given the [frequent coverage](https://bitcoinops.org/en/topics/consensus-cleanup-soft-fork/) of Antoine’s [ongoing research](https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/5) on reviving the Great Consensus Cleanup that includes a similar fix. It seems better to me to package all four fixes in one soft fork. Was it overlooked that this work is being duplicated, or are you disagreeing with the direction of Antoine’s work and want to put forward an alternative? Could you elaborate why you decided to submit this?

HI Murch! 

>Great Consensus Cleanup

While the GCC is a great name for a disjoint set of bitcoin protocol security fixes, that doesn't mean they all need to be bundled into one BIP and deployed at the same time.

>It seems better to me to package all four fixes in one soft fork.

This document takes no opinion on how this enhancement is deployed.

Previously we have deployed multiple consensus changes in a single soft fork that contained multiple BIPs. The 2017 Segwit soft fork had [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki), [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki). The Taproot soft fork had [BIP341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki) and [BIP342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki).

>Was it overlooked that this work is being duplicated, 

I was recommended that this was a piece of work that was easily to peel off from the rest of GCC and propose a fix for. Its relatively simple and straight forward, and has been known about for a long time.

> Antoine's on going research.

I think Antoine's research is great, but just because he is researching the topic doesn't mean he gets to monopolize the vulnerabilities. The 64 byte transaction vulnerability has been known since at least 2012 - yet no one has taken the time to write up a BIP to fix it 😬 . I've put forth the first BIP draft in the 13 years since this vulnerability has been known. I encourage Antoine to write up his proposal (hopefully sooner rather than later) as he doesn't seem to be interested in working with me. 

Other vulnerabilities in GCC have also been known about for a long time - such as worse case block validation time. Antoine has done a lot of novel research there and I think his work product has been excellent on delving. 

There is plenty of meat left on the GCC bone, this was a small piece I decided to carve off.

murchandamus commented on bip-XXXX.mediawiki:
The area around "both the coinbase transaction" seems to be missing something

murchandamus commented on bip-XXXX.mediawiki:
Could you explain the exact semantics of how and where such transactions are disallowed?

Do you mean e.g., that a block including a 64-byte transaction must be rejected as invalid?

murchandamus commented on bip-XXXX.mediawiki:
Could you expand on how you are suggesting that implementers should mitigate this incompatibility? What about the potential existence of a pre-signed transaction that serializes to 64 bytes?

murchandamus commented on bip-XXXX.mediawiki:
A lot of the content of this section would maybe fit better under Motivation. Rationale usually explores the design decisions, alternate designs, and related work. For example you could expand on the approach of tracking the height of the Merkle branch to the coinbase transaction in the Rationale.

murchandamus reviewed (commented):
This does seem like a reasonable start for a BIP. A few parts of the content could perhaps be organized a bit differently, and some sections could be expanded. Especially the Specification could be a bit more precise in what an implementation should do to be compliance with this BIP. I left you a few editorial nits that you could perhaps take a look at.

It seems likely to me that this submission may be duplicating some effort on all of our ends, assuming that the other anticipated BIP gets submitted near-term, but fair enough, we have not received another BIP on this topic. I guess I can’t fault multiple people for being interested in the same topic.

Christewart commented:
I believe the CI failure is unrelated to me?

jonatack commented:
> I believe the CI failure is unrelated to me?

Yes, it will go away when you next rebase to current master.

jonatack commented on bip-XXXX.mediawiki:
s/its/their/, or "transaction" in the singular form

```suggestion
This BIP disallows bitcoin transactions that are serialized to 64 bytes in length without their witness.
```

Christewart replied:
https://github.com/bitcoin/bips/pull/1760#discussion_r1987577059

jonatack commented on bip-XXXX.mediawiki:
```suggestion
This has been prevented by relay policy and the RPC interface since 2018<ref>[https://github.com/bitcoin/bitcoin/pull/11423/commits/7485488e907e236133a016ba7064c89bf9ab6da3 PR #11423 disallows 64-byte transactions in Bitcoin Core relay and RPC interface]</ref>.
```

jonatack commented on bip-XXXX.mediawiki:
IIUC
```suggestion
and add the block to their blockchain. Peers that receive the invalid block before the valid block
```

jonatack commented on bip-XXXX.mediawiki:
"Finding this collision approximately 22 bits worth of work" -> looks like a verb is missing here, or the sentence seems hard to follow

jonatack commented on bip-XXXX.mediawiki:
```suggestion
for this type of transaction is ephemeral anchor outputs.<ref>[https://bitcoinops.org/en/topics/ephemeral-anchors/ What are ephemeral anchor outputs? This allows anyone on the network to use that output as the input to a child transaction. This allows anyone to create the fee-paying child, even if they don’t receive any of the other outputs from the parent transaction. This allows ephemeral anchors to function as fee sponsorship but without requiring any consensus changes.]</ref>
```

jonatack commented on bip-XXXX.mediawiki:
```suggestion
Pre-segwit 64-byte transactions that spend a nonstandard utxo that are inherently malleable.<ref>[https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#trust-free-unconfirmed-transaction-dependency-chain Details on how to malleate a pre-segwit transaction]</ref>
```

jonatack commented on bip-XXXX.mediawiki:
```suggestion
with the last transaction 7f2efc6546011ad3227b2da678be0d30c7f4b08e2ce57b5edadd437f9e27a612<ref>[https://mempool.space/tx/7f2efc6546011ad3227b2da678be0d30c7f4b08e2ce57b5edadd437f9e27a612 Last 64-byte transaction in the bitcoin blockchain]</ref>
```

jonatack commented on bip-XXXX.mediawiki:
```suggestion
 * malleability checks that may interact with other CheckBlock checks.
```

Christewart replied:
I'm going to avoid changing these nits unless they change in the reference implementation unless you have a super strong opinion on this.

jonatack replied:
I see. It wasn't clear to me that this is extracted from the open pull in bitcoin-inquisition.

jonatack commented on bip-XXXX.mediawiki:
```suggestion
 * Note that this is called before merkle-tree checks and so must never return a
```

jonatack commented on bip-XXXX.mediawiki:
```diff
+This sample implementation is currently an open pull request here:
+
```

https://github.com/bitcoin-inquisition/bitcoin/pull/24/files

jonatack commented on bip-XXXX.mediawiki:
Now that "Bitcoin Core's" was added.

```suggestion
This has been prevented by Bitcoin Core's relay policy and RPC interface since 2018<ref>[https://github.com/bitcoin/bitcoin/pull/11423/commits/7485488e907e236133a016ba7064c89bf9ab6da3 PR #11423 disallows 64-byte transactions in Bitcoin Core relay and RPC interface]</ref>.
```

jonatack replied:
Perhaps also mention https://github.com/bitcoin/bitcoin/commit/8c5b3646b5afe8a61f5c66478d8e11f0d2ce5108 here

jonatack commented on bip-XXXX.mediawiki:
```diff
@@ -0,0 +1,160 @@
+
+==Abstract==
+
+This BIP describes the rationale for disallowing transactions that are serialized to 64 bytes without the transaction's witness. 
```
nit, extra space at EOL
```suggestion
This BIP describes the rationale for disallowing transactions that are serialized to 64 bytes without the transaction's witness.
```

jonatack commented on bip-XXXX.mediawiki:
nit, extra space at EOL

```suggestion
The peer receiving the malicious block marks the block as invalid as T<sub>m</sub>
```

Goddessnunu22 commented:
> This BIP makes 64 byte bitcoin transactions serialized without the witness consensus invalid
> 
> Mailing list post: https://groups.google.com/g/bitcoindev/c/rf3QOlzg230/m/eOOC8pkOAgAJ
> 
> Delving discussion: https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710
> 
> Side note, wasn't sure how to best handle the 2-Bitcoin-Merkle.pdf as it isn't hosted anywhere easily accessible AFAICT? For now I just added it into the git repo. Lmk if you have other suggestions

@Alicia Prince

jonatack commented:
Would be good to have some more eyes on this, otherwise seems close to ready as a draft for a number assignment.

vostrnad commented:
@Christewart Antoine just opened a pull request for the Consensus Cleanup which includes the 64-byte transaction fix: #1800

If you don't feel too strongly about this being a separate BIP, I think it would be better to withdraw it, otherwise it'll just be duplicating effort with no obvious benefit.

Christewart commented:
Hi @vostrnad,

As of [this commit](https://github.com/bitcoin/bips/pull/1800/commits/098894f04f41494b0329a3e0479e27f33f841812), PR #1800 does not provide as much detail as this document in defining the problem with 64-byte transactions, explaining the proposed solution, and demonstrating why it does not pose a confiscatory risk.

I hope the author of #1800 will consider expanding the BIP with more detail or breaking it into separate proposals for each consensus change. However, based on [this email](https://groups.google.com/g/bitcoindev/c/0tSvml90Qcw/m/2GAFUCRmBgAJ), it seems they believe the current content is sufficient and only requires minor edits.

At this time, I believe my document is of higher quality. I will not be voluntarily closing my BIP until I feel #1800 reaches parity.

murchandamus commented:
Let’s call this BIP 53. Please update the table entry preamble and title correspondingly.

Christewart commented:
> Let’s call this BIP 53. Please update the table entry preamble and title correspondingly.

Done in 23e01bf

murchandamus commented on bip-XXXX.mediawiki:
I think "non-standard" is more common in the context of transactions. Could you use either "non-standard" or "nonstandard" throughout the document?

```suggestion
There are 6<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/nonstandard-hashlock-utxos.txt As of block `00000000000000000001194ae6be942619bf61aa70822b9643d01c1a441bf2b7` there exist 6 non-standard hashlock utxos that could theoretically have a 0-3 byte pre-image. None of them have a 0-3 byte pre-image.]</ref>
non-standard hashlock utxos in the bitcoin blockchain. None of them have a 0-3 byte pre-image. This means they cannot be spent by a 64-byte transaction.

Pre-segwit 64-byte transactions that spend a non-standard utxo that are inherently malleable.<ref>[https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#trust-free-unconfirmed-transaction-dependency-chain Details on how to malleate a pre-segwit transaction]</ref>
Policy rules such as CLEANSTACK, MINIMALDATA, PUSHONLY are not consensus rules. If a user has a way to confirm an already non-standard
```

murchandamus commented on bip-XXXX.mediawiki:
```suggestion
a workable second transaction which will hash to the second 32 bytes)—a very
```

murchandamus commented on bip-XXXX.mediawiki:
Consistency-nit: Please stick either to "Merkle" or to "merkle" throughout the document.

murchandamus reviewed (commented):
The document seems fairly complete content-wise, but the presentation could use more attention. Capitalization is a bit random: some acronyms are capitalized ("SPV"), but others are uncapitalized (e.g. "utxos") and proper nouns are sometimes capitalized (Merkle, Bitcoin Core), sometimes not (merkle, bitcoin). Please either retitle one of the two Rationale sections or combine them.

It would be good if the proposal got more commentary or review from other contributors, but otherwise seems close to ready for a merge.

murchandamus commented:
Please also add the table entry to the README.

murchandamus commented on bip-0053.mediawiki:
```diff
@@ -6,7 +6,7 @@
   Status: Draft
   Type: Standards Track
-  Created: ?
+  Created: 2024-12-20
```
According to [BIP2](https://github.com/bitcoin/bips/blob/master/bip-0002.mediawiki#bip-header-preamble), the "Created" header records the date a proposal was assigned a number, which is why I suggested 2025-04-11.

![image](https://github.com/user-attachments/assets/5ac62de5-00b6-455e-940f-f22c88d228e1)

murchandamus reviewed (commented):
I assume you are still working on my other suggestions, so I’ll take another look later.

Christewart commented:
Yes I am. Child nap time only allows for so many things to get done at 1
time 😅. I’ll try to address the rest later today

On Tue, Apr 22, 2025 at 12:12 PM murchandamus ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> I assume you are still working on my other suggestions, so I’ll take
> another look later.
> ------------------------------
>
> In bip-0053.mediawiki
> <https://github.com/bitcoin/bips/pull/1760#discussion_r2054517864>:
>
> > @@ -6,7 +6,7 @@
>    Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0053
>    Status: Draft
>    Type: Standards Track
> -  Created: ?
> +  Created: 2024-12-20
>
> According to BIP2
> <https://github.com/bitcoin/bips/blob/master/bip-0002.mediawiki#bip-header-preamble>,
> the "Created" header records the date a proposal was assigned a number,
> which is why I suggested 2025-04-11.
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/5ac62de5-00b6-455e-940f-f22c88d228e1>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bips/pull/1760#pullrequestreview-2784772967>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA22ETNRE5UYZ32HHFCPFTD22ZZ6DAVCNFSM6AAAAABW3SF3JSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDOOBUG43TEOJWG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Christewart commented:
> The document seems fairly complete content-wise, but the presentation could use more attention. Capitalization is a bit random: some acronyms are capitalized ("SPV"), but others are uncapitalized (e.g. "utxos") and proper nouns are sometimes capitalized (Merkle, Bitcoin Core), sometimes not (merkle, bitcoin). 

Ok think this is addressed in [f797b9f](https://github.com/bitcoin/bips/pull/1760/commits/f797b9f5ef2ed7f0ef91985cf8a4f7c0970e2119), [bd815e9](https://github.com/bitcoin/bips/pull/1760/commits/bd815e97e7fb9741596e8bfdd4320c80ec1e487e), [acda401](https://github.com/bitcoin/bips/pull/1760/commits/acda401f9c4bd150dfb082e1356e4d6c646f6433)

>Please either retitle one of the two Rationale sections or combine them.

Sorry this was a bit unclear could you be more specific of what you would like to see?

murchandamus commented:
> Also I missed some code review due to github.com collapsing some comments, hopefully everything is addressed now with [5a89036](https://github.com/bitcoin/bips/commit/5a890366db2f8cf4a3a2a4c38e1c65590d45e6ef)

Yeah, I think I see what happened: when you started by moving the file, it marked all of my review comments as outdated, because they had been left on the old file. ;)

murchandamus reviewed (commented):
Reviewed the changes, looks much better, thanks. :)

What’s your plan for this proposal? Do you see it as a standalone soft fork that can be merged as is on its own, or are you working on additional parts? What is the intended deployment strategy?

Christewart commented:
>Do you see it as a standalone soft fork that can be merged as is on its own

This is possible if the community wants to do this.

>or are you working on additional parts? 

I am not at this time.

>What is the intended deployment strategy?

I don't have one at this time. Recent precedent with [BIP148](https://github.com/bitcoin/bips/blob/master/bip-0148.mediawiki) and [BIP343](https://github.com/bitcoin/bips/blob/master/bip-0343.mediawiki) details activating the last 2 soft forks that contained a variety of BIPs. 

These activation BIPs were written by a different set of people than the BIP authors for taproot and segwit. I assume something like this will have to happen again as it seems like there isn't really another realistic path to deploying soft forks IMO.

murchandamus commented on bip-0053.mediawiki:
```suggestion
Suhas Daftuar, AJ Towns, Sergio Demian Lerner, Greg Maxwell, Matt Corallo, Antoine Poinsot, Dave Harding and Eric Voskuil
```

murchandamus reviewed (commented):
Okay, is there anything else that is missing from your perspective before this can be merged?

Christewart commented:
> Okay, is there anything else that is missing from your perspective before this can be merged?

I thought about this a bit more, there are 2 things that could potentially be addressed. I would like to hear your opinion.

How BIP3 defines the `Rationale` section:

>The rationale should record relevant objections or important concerns that were raised and addressed as this proposal was developed.

One concern that was raised on the [mailing list by Jeremy Rubin](https://groups.google.com/g/bitcoindev/c/0tSvml90Qcw/m/M8NRIlFqBgAJ?utm_medium=email&utm_source=footer) and agreed by Eric Voskuil was that disallowing 64 byte transaction 

>"might integrate very poorly with this kind of edge condition"

Do you think this is adequately addressed by [this](https://github.com/bitcoin/bips/blob/fd3bc98ba467ffbdb6e1df463a3cf9e64d0781bf/bip-0053.mediawiki#user-content-Segwit_64byte_transactions) section of the BIP? I'm on the fence. 

One other concern shared on the [delving thread by Antoine Poinsot](https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/84?u=chris_stewart_5) was sidechains like Citrea may be vulnerable to the same merkle tree weakeness that SPV clients are. I haven't investigated this further, but it could potentially merit a section the `Motivation`.

This currently isn't at the top of my bitcoin research priorities, so if you think these things should be done I'll address them next week.

Christewart commented:
>  we can also merge this PR as a draft and you can continue to iterate on it as you move it toward "Proposed".

Lets do that. I rebased the PR to get rid of conflicts. Would you like me to squash everything into 1 commit as well?

murchandamus commented:
@Christewart: Yeah, I think it would make sense to squash it. Thanks.

murchandamus commented on README.mediawiki:
```diff
@@ -301,6 +301,14 @@ Those proposing changes should consider that ultimately consent may rest with th
+| Chris Stewart
+| Standard
+| Draft
+<!-- 50 series reserved for a group of post-mortems -->
```
this line seems outdated and if anywhere, it should be after BIP 50 or after BIP 54.

```suggestion
```

Christewart commented:
> @Christewart: Yeah, I think it would make sense to squash it. Thanks.

Done in c07e507

jonatack commented on bip-0053.mediawiki:
```suggestion
BIP37<ref>[https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki BIP37]</ref>provides a partial Merkle tree format<ref>[https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki#partial-merkle-branch-format Partial Merkle Tree Format]</ref>
```

jonatack commented on bip-0053.mediawiki:
Link is 404

jonatack commented on bip-0053.mediawiki:
computationally

jonatack commented on bip-0053.mediawiki:
```suggestion
of SPV proofs by 70%.<ref>[https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/29 Base proof: 80-byte header + 448-byte partial Merkle tree = 528 bytes. Proof with coinbase tx, assuming the coinbase tx is in the left half of the tree and the tx to prove is in the right half of the tree: 80-byte header + 416 bytes partial Merkle tree for coinbase tx + 416 bytes partial Merkle tree for tx = 912 bytes.]</ref>
```

jonatack commented on bip-0053.mediawiki:
- Ambiguous "This" here ("This function"?) following "This" in the preceding sentence that appears to refer to a different subject

- Suggest specifying the implementation (presumably Bitcoin Core) and version along with "currently"

Christewart replied:
Same response as here: https://github.com/bitcoin/bips/pull/1760#discussion_r1987578520

jonatack commented on bip-0053.mediawiki:
```suggestion
 * Note that as this function is called before merkle-tree checks, it must never return a
```

Christewart commented on bip-0053.mediawiki:
I don't think [#26265](https://github.com/bitcoin/bitcoin/pull/26265) got merged so it doesn't really make sense to link to it?

Christewart commented on bip-0053.mediawiki:
How should i be linking these? These are the 2 options I see

1. Its my understanding I cannot use relative paths inside of mediawiki (i.e. `../bip-0053/2-BitcoinMerkle.pdf`
2. If I link to a commit, the commit will likely be squashed and no longer exist after the rebase.

Christewart replied:
I fixed them by just using this link, seems like the problem occurred when we renamed from bip-xxxx -> bip-0053

https://github.com/Christewart/bips/tree/2024-12-20-64bytetxs/bip-0053

Christewart commented:
@jonatack I believe every non-nit comment has been addressed

murchandamus commented on bip-0053.mediawiki:
```diff
@@ -64,7 +64,7 @@ This attack vector was fixed in 0.6.2<ref>[https://bitcoin.org/en/alert/2012-05-
 
 Producing a valid Bitcoin transaction T<sub>m</sub> that adheres to network consensus
-rules requires 224 bits of work<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-XXXX/2-BitcoinMerkle.pdf Note that the first transaction in a block must be a coinbase, and as discussed
+rules requires 224 bits of work<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-0053/2-BitcoinMerkle.pdf Note that the first transaction in a block must be a coinbase, and as discussed
```
```suggestion
rules requires 224 bits of work<ref>[bip-0053/2-BitcoinMerkle.pdf Note that the first transaction in a block must be a coinbase, and as discussed
```

murchandamus commented on bip-0053.mediawiki:
```diff
@@ -51,7 +51,7 @@ is not a valid transaction according to network consensus rules.
 and add the block to their blockchain. Peers that receive the invalid block before the valid block
 will never come to consensus with their peers due to the malicious user finding a collision
-within the block's Merkle root. Finding this collision is approximately 22 bits worth of work.<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-XXXX/2-BitcoinMerkle.pdf to produce a block having a Merkle root that
+within the block's Merkle root. Finding this collision is approximately 22 bits worth of work.<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-0053/2-BitcoinMerkle.pdf to produce a block having a Merkle root that
```
While those links work at the time, a relative link would make more sense to me, because it would work now, and still link to the correct document if it gets updated later.

```suggestion
within the block's Merkle root. Finding this collision is approximately 22 bits worth of work.<ref>[bip-0053/2-BitcoinMerkle.pdf to produce a block having a Merkle root that
```

murchandamus commented on bip-0053.mediawiki:
```diff
@@ -82,7 +82,7 @@ Notably this format does not commit to the height of the Merkle tree.
 are less constrained than the first 32 bytes) are constructed so that they collide
 with the hash of some other fake, invalid transaction F. The attacker can fool the SPV client into believing that F
-was included in a Bitcoin block rather than T with 81 bits<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-XXXX/2-BitcoinMerkle.pdf An attacker who can do 81 bits of work (followed by another 40 bits of work, to
+was included in a Bitcoin block rather than T with 81 bits<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-0053/2-BitcoinMerkle.pdf An attacker who can do 81 bits of work (followed by another 40 bits of work, to
```
```suggestion
was included in a Bitcoin block rather than T with 81 bits<ref>[bip-0053/2-BitcoinMerkle.pdf An attacker who can do 81 bits of work (followed by another 40 bits of work, to
```

murchandamus commented on bip-0053.mediawiki:
```diff
@@ -98,7 +98,7 @@ of SPV proofs by 70%.<ref>[https://delvingbitcoin.org/t/great-consensus-cleanup-
 
 There have been 5 64-byte transactions that have occurred in the Bitcoin blockchain as of this
-writing <ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/64byte-tx-mainnet.txt 64-byte transactions in the Bitcoin blockchain]</ref>
+writing <ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-0053/64byte-tx-mainnet.txt 64-byte transactions in the Bitcoin blockchain]</ref>
```
```suggestion
writing <ref>[bip-0053/64byte-tx-mainnet.txt 64-byte transactions in the Bitcoin blockchain]</ref>

murchandamus commented on bip-0053.mediawiki:
```diff
@@ -107,7 +107,7 @@ occurring at block height 419,606<ref>[https://mempool.space/block/0000000000000
 The largest scriptSig a pre-segwit 64-byte transaction can have is 4 bytes.<ref>[https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/73]</ref>
 
-There are 6<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/non-standard-hashlock-utxos.txt As of block `00000000000000000001194ae6be942619bf61aa70822b9643d01c1a441bf2b7` there exist 6 non-standard hashlock UTXOs that could theoretically have a 0-3 byte pre-image. None of them have a 0-3 byte pre-image.]</ref>
+There are 6<ref>[https://github.com/Christewart/bips/blob/2024-12-20-64bytetxs/bip-0053/non-standard-hashlock-utxos.txt As of block `00000000000000000001194ae6be942619bf61aa70822b9643d01c1a441bf2b7` there exist 6 non-standard hashlock UTXOs that could theoretically have a 0-3 byte pre-image. None of them have a 0-3 byte pre-image.]</ref>
```
```suggestion
There are 6<ref>[bip-0053/non-standard-hashlock-utxos.txt As of block `00000000000000000001194ae6be942619bf61aa70822b9643d01c1a441bf2b7` there exist 6 non-standard hashlock UTXOs that could theoretically have a 0-3 byte pre-image. None of them have a 0-3 byte pre-image.]</ref>
```

murchandamus reviewed (commented):
Most of these changes look good, but it seems better to use a relative link for the accompanying files than a external link to the current version of the draft.

Christewart commented:
> Most of these changes look good, but it seems better to use a relative link for the accompanying files than a external link to the current version of the draft.

Unfortunately this doesn't seem to work? I've accepted one of your commits you suggested ( bc9afbc ) and here is the result: https://github.com/bitcoin/bips/blob/bc9afbca79d484970756aa92659d246c049fbb55/bip-0053.mediawiki#cite_note-7

murchandamus commented on bip-0053.mediawiki:
Sorry, for the incorrect suggestions. You will want to use [internal link formatting](https://www.mediawiki.org/wiki/Help:Links#Internal_links), which separates link and text with a pipe and encloses in double brackets:
```suggestion
rules requires 224 bits of work<ref>[[bip-0053/2-BitcoinMerkle.pdf|Note that the first transaction in a block must be a coinbase, and as discussed
above, that largely constrains the first 32 bytes of the first transaction: only
the 4 version bytes are unconstrained. So it would take at least 28*8= 224 bits
of work to find the first node in a given row of the tree that would match the
first half of a coinbase, in addition to the amount of work required to grind the
second half of the transaction to something meaningful (which is much easier –
only 16 bytes or so are constrained, so approximately 128 bits of work to find a collision). Of course, any of the rows in the Merkle tree could be used, but it nevertheless seems clear that this should be computationally infeasible.]]</ref>.
```

I tested that, and it works: 
![image](https://github.com/user-attachments/assets/3b15b7db-99c6-4e15-9685-bbe508da2f17)

murchandamus replied:
I realized that after figuring out the fix, I might as well just commit it. ;)
I tested it, it works. :)

murchandamus commented on bip-0053.mediawiki:
```suggestion
was included in a Bitcoin block rather than T with 81 bits<ref>[[bip-0053/2-BitcoinMerkle.pdf|An attacker who can do 81 bits of work (followed by another 40 bits of work, to
construct the funding transaction whose coins will be spent by this one) is able
to fool an SPV client in this way.]]</ref> of work. This also reduces implementation complexity for SPV wallets<ref>[https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710/43 The steps needed to make sure a Merkle proof for a transaction is secure.]</ref>.
```

murchandamus commented on bip-0053.mediawiki:
```suggestion
within the block's Merkle root. Finding this collision is approximately 22 bits worth of work.<ref>[[bip-0053/2-BitcoinMerkle.pdf|to produce a block having a Merkle root that
is a hash of a 64-byte quantity that deserializes validly, it’s enough
to just do 8 bits of work to find a workable coinbase (which will hash to the first
32 bytes), plus another ≈22 bits of work ((1/5) ∗224, so slightly less) to find
a workable second transaction that will hash to the second 32 bytes)– a very
small amount of computation.]]</ref>
```

murchandamus reviewed (commented):
Sorry for the incorrect suggestions, you will need to use internal links instead of external links to use relative paths. I updated my suggestions.

murchandamus commented:
I realized that after figuring out the fix for the links, I might as well commit it. All the links to the documents shipping with the BIP are now internal relative links, and I tested that they work. :)

murchandamus commented:
Squashed per request of @Christewart.

Christewart commented:
Ok @murchandamus @jonatack I believe i've incorporated all outstanding review or added a comment on why I'm not taking a suggestion. I squashed everything again into 1 commit
